### PR TITLE
[BLU-3628] Simplify access management

### PR DESCRIPTION
**Pull request by lazaralex98**
- removed the `Ownable` inheritance in favor of using the `DEFAULT_ADMIN_ROLE` in `AccessControl`
- assigned the `DEFAULT_ADMIN_ROLE` to the deployer in the constructor using `_grantRole` as `_setupRole` is deprecated
- removed custom role management in favor of `AccessControl` based roles (for pauser and token issuer)

### CODE REVIEW

**Comment by aspiers** [comment]
@lazaralex98 

```
> ethers.constants.HashZero  // DEFAULT_ADMIN_ROLE
'0x0000000000000000000000000000000000000000000000000000000000000000'
> ethers.utils.id('MANAGER_ROLE')
'0x241ecf16d79d0f8dbfb92cbc07fe17840425976cf0667f022fe9877caa831b08'
```

Also remember you can comment directly on specific lines of your own PR, which is sometimes a better place to put questions about specific bits of code than in the PR description or PR-level comments.

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:42]: Near the top of the file:

```typescript
    import { utils } from 'ethers';
```

and then

```suggestion
		await bridge.connect(admin).grantRole(utils.id("TOKEN_ISSUER_ROLE"), tokenIssuer.address);
```

although I'm pretty sure you can also do:

```typescript
		const issuerRole = await bridge.TOKEN_ISSUER_ROLE();
```

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:47]: No, this is wrong.  `DEFAULT_ADMIN_ROLE` is the zero value.  Put this after the imports:

```typescript
    const DEFAULT_ADMIN_ROLE = ethers.constants.HashZero;
```

then

```suggestion
		const isAdmin = await bridge.hasRole(DEFAULT_ADMIN_ROLE, admin.address);
```

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:64]: Same here, use `utils.id`.

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:7]: The presence of `lib` here looks wrong to me - seems to be assuming knowledge of `ethers` internals.

I would suggest

```javascript
const { utils } = require("ethers");
const { id } = utils;
```

or just drop the second line and use `utils.id`.

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:255]: This would not be a helpful error message because it displays a hash.

```suggestion
				`AccessControl: account ${broker.address.toLowerCase()} is missing TOKEN_ISSUER_ROLE`
```

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:330]: Same here and below.

File: test/ToucanRegenBridge.js
Comment (aspiers) [inline test/ToucanRegenBridge.js:147]: Good catch but how did this pass previously?

Comment (lazaralex98) [reply test/ToucanRegenBridge.js:147]: That is an excellent question that I couldn't answer myself. The error was, for whatever reason, introduced in https://github.com/regen-network/toucan-bridge/pull/61/commits/2d63ffef795a3ac197bdb6d500f5e547606ec247 and wasn't an issue previously.

But it's bizarre because there is no specific place where a `broker` becomes a token issuer before that, not in the contract or test. The flow previously is: the contract gets deployed with `bridgeAdmin` (now `tokenIssuer`) as a token issuer. At one [later point](https://github.com/regen-network/toucan-bridge/blob/master/test/ToucanRegenBridge.js#L326), the token issuer role is given to `admin`. But never to `broker`.

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [inline contracts/ToucanRegenBridge.sol:3]: Why is this change needed?

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [inline contracts/ToucanRegenBridge.sol:91]: Since we are going to assign these roles to other accounts as soon as the contract is deployed, we may as well provide the correct accounts as part of the constructor inputs.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:3]: Because I was getting the following error:
`Source file requires different compiler version
(current compiler is 0.8.17+commit.8df45f5f.Emscripten.clang)
- note that nightly builds are considered to be strictly
less than the released version`

Comment (0xmichalis) [reply contracts/ToucanRegenBridge.sol:3]: What process is returning this error? The compiler version should be driven by the [HH config](https://github.com/regen-network/toucan-bridge/blob/7eec01c88e2c26cdda7abc84c2eddc4a61b4bfc0/hardhat.config.js#L46).

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:91]: 2 arrays, one with addresses and one with roles + a loop over them to assign roles. That's the design I'm thinking of.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:3]: My IDE was flagging this as an issue and I couldn't benefit from "format on save" and stuff like that, we can delete the commit that introduces this after we're done with the PR if you want

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:6]: Opps, my bad.

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [reply contracts/ToucanRegenBridge.sol:3]: We definitely need to revert this change as it has been deliberately pinned to a fixed version so it's clearer to auditors with what version are we deploying.

Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:3]: Makes sense to me, I'll revert it after refinement today.

Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:3]: Actually, I'll rewrite history to remove it instead of reverting it.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:95]: I actually feel more comfortable keeping this in its own commit, I'm not sure in which other commit it would make sense to add it

File: contracts/ToucanRegenBridge.sol
Comment (aspiers) [inline contracts/ToucanRegenBridge.sol:22]: This commit's changes look good, but again the commit message could use improvement.

Most importantly (as always) it should contain an explanation of *why* this change is being done.  It could mention that we are doing this to unify access control into a single mechanism, and include the reasons our auditors gave us for why this is beneficial.

It could also explain why this change allows us remove an event and an existing function.

Finally, either this commit should come before the variable rename commit (which is currently the one before it), otherwise the rename is premature, or actually there is an argument that the two commits could be squashed together because they are strongly linked.

File: test/ToucanRegenBridge.js
Comment (aspiers) [reply test/ToucanRegenBridge.js:146]: I agree this seems odd, and thanks for the analysis.

2d63ffe is no longer part of this PR; however I see that 22536b8f (which *is* in the PR) is essentially identical, except for doing `require(hasRole(...))` instead of `onlyRole(...)`.

Does 22536b8f introduce the same error?  If not, then you could potentially compare them to figure out where the error crept in, although maybe it's not worth the effort given the urgency to get other stuff done.

However if it does then this means an error was introduced in a previous commit in this PR, and then fixed here in this commit - if so, then as a reminder that would be going against part of [our coding standards](https://github.com/ToucanProtocol/tokenizer/blob/master/documentation/coding-standards.md):

image

It's totally fine if we can't fully explain why the breakage happened, as long as we explicitly call that out and make sure the above golden rule wasn't broken.  If it was (i.e. if 22536b8f introduced the same error as 2d63ffe), then the simple solution would be to squash these two commits together, and explain the uncertainty in the commit message.  The information you gave in the comment above is already very helpful, so you could just include that in the combined commit message.

File: contracts/ToucanRegenBridge.sol
Comment (aspiers) [inline contracts/ToucanRegenBridge.sol:93]: Any particular reason for this requirement?  If so, please include it in the commit message.

But I can imagine situations where it's useful to deploy and then set up the roles later, e.g. if we hadn't yet decided who should get which roles, or if EOAs or multisigs still needed to be set up.

File: contracts/ToucanRegenBridge.sol
Comment (aspiers) [inline contracts/ToucanRegenBridge.sol:101]: `in the arrays` is redundant here and just bloats the contract size.

**Review by aspiers — COMMENTED** [review]
Except for the nitpick about the error message, the changes here all LGTM - my only feedback is about the commit messages and the commit ordering/grouping.  If that feedback, or at least most of it, can be addressed in under an hour (which I think should be possible) then please do it, but basically these changes have my approval and as long as @0xmichalis is also OK with them then I would like to see them merged ASAP so that we can proceed with redeploy on mumbai on Monday, since this is urgent.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:22]: I'd rather have them separate, tbh. It's easier to distinguish what happened this way. 

I also think it makes sense for the `bridgeAdmin -> tokenIssuer` rename to occur before this because that was a refactor that made it easier for me to know who's who when modifying the tests in the commit that adds `TOKEN_ISSUER_ROLE`.

That said, I can see how the messages could describe all of this better.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:93]: Wouldn't the `for` loop fail if it tried to iterate over an array of length 0?

**Comment by lazaralex98** [comment]
https://github.com/regen-network/toucan-bridge/pull/61#discussion_r1118130106

Agree about squashing them. I do feel that keeping them separate made the review easier.

File: test/ToucanRegenBridge.js
Comment (lazaralex98) [reply test/ToucanRegenBridge.js:146]: Agree about squashing them. I do feel that keeping them separate made the review easier.

File: contracts/ToucanRegenBridge.sol
Comment (aspiers) [reply contracts/ToucanRegenBridge.sol:93]: I don't see why it would - what do you mean by fail?  The termination condition is `i < accounts.length`, so if `accounts.length` is `0` then it will just terminate immediately, which is what we want.

Comment (aspiers) [reply contracts/ToucanRegenBridge.sol:93]: I'm 99.9% sure but you should make sure yourself along the "trust but verify" maxim of  blockchain ;-)

The way to check is to write test cases for constructing with no roles and with one or more roles.

Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:93]: You were correct. I don't know why I had this idea in my mind that it would fail. I'll remove the line requiring the `length > 0`. It might make sense to add the test though.

**Review by aspiers — APPROVED** [review]
LGTM now, thanks!  Whoever has maintainer access to this repo, please can you merge?

**Comment by lazaralex98** [comment]
@0xmichalis seems to be the person.

**Comment by 0xmichalis** [comment]
@ryanchristo can you please grant @aspiers @bezerrablockchain with access to merge PRs? Thanks! I think it's also @pheuberger who should be able to merge.

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [reply contracts/ToucanRegenBridge.sol:95]: We should avoid granting access to `msg.sender` in case my original comment wasn't clear and only assign to the provided accounts.

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [inline contracts/ToucanRegenBridge.sol:99]: We can avoid assigning roles to `msg.sender` since post-deployments all of the roles need to be revoked and avoid restricting `DEFAULT_ADMN_ROLE` here. Actually this check should be removed altogether and we should just ensure there is a single admin assigned.

**Comment by clevinson** [comment]
@bezerrablockchain and @pheuberger already had "write" access, which should give them ability to merge. I've added @aspiers now as well.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:99]: Hmm, similar to what Rodrigo has done in https://github.com/ToucanProtocol/tokenizer/pull/2234/files#diff-c68e1476ab8eae3300691c9701dbb9ac5fdaddc28023a8753e63ef3728a80f3f, but in this case, we want to ensure there is **just one** admin.

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [inline contracts/ToucanRegenBridge.sol:100]: Probably better not to restrict this to one admin
```suggestion
        require(adminAccounts >= 1, "should have at least one admin");
```
Could even simplify `adminAccounts` to a boolean

File: contracts/ToucanRegenBridge.sol
Comment (0xmichalis) [inline contracts/ToucanRegenBridge.sol:93]: Since there is no setter for `nctPool`, we should add a `require` here to ensure the pool is non-zero. Not related to your current changes but probably best to have in a follow-up PR before we deploy.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:100]: > Probably better not to restrict this to one admin

That's okay, I got the idea from your previous comment that we only want a single admin, but I can easily modify this. Thanks for the clarification.

File: contracts/ToucanRegenBridge.sol
Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:93]: I'll open a quick issue to keep track of it.

Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:93]: https://linear.app/toucan/issue/BLU-3756

Comment (lazaralex98) [reply contracts/ToucanRegenBridge.sol:93]: Also a PR https://github.com/regen-network/toucan-bridge/pull/62

**Review by 0xmichalis — APPROVED** [review]
LGTM, good work!

**Pull request merged**
